src/react/use-api-mutation/index.tsx

```diff
@@ -0,0 +1,101 @@
+import {useContext, useState} from 'react'
+
+import {Api} from '../../api'
+import {
+  ApiRequestMethod,
+  ApiRequestOptions,
+  ApiRequestParams,
+  ResponseBody
+} from '../../api/typings'
+import {ApiContext} from '../context'
+import {
+  MutationApiHelpers,
+  UseApiMutationParams,
+  UseApiMutationReturnFunction,
+  UseApiMutationReturnValue
+} from './typings'
+
+export * from './typings'
+
+export function useApiMutation<
+  TMutationArgs extends any[],
+  TMutationReturnValue
+>(
+  params: UseApiMutationParams<TMutationArgs, TMutationReturnValue>
+): UseApiMutationReturnValue<TMutationArgs> {
+  const mutationApiHelpers = useMutationApiHelpers()
+  const [mutating, {registerMutation, deregisterMutation}] = useIsMutating()
+
+  const mutate: UseApiMutationReturnFunction<TMutationArgs> = async (
+    ...mutationArgs
+  ) => {
+    registerMutation()
+
+    const mutator = params.mutation(...mutationArgs)
+    let returnValue: TMutationReturnValue
+
+    try {
+      returnValue = await mutator(mutationApiHelpers)
+    } catch (error) {
+      if (params.onError) {
+        return params.onError(error, {mutationArgs})
+      }
+      throw error
+    } finally {
+      deregisterMutation()
+    }
+
+    params.onSuccess?.(returnValue, {mutationArgs})
```

Comment:
Is there a reason this needs to run after the try-catch block?

Reply:
This way, if an error occurs inside the `onSuccess` handler, `onError` will not trigger. This could avoid situations where a success snackbar shows via `onSuccess`, but then an error occurs after that, causing `onError` to be called and immediately triggering an error snackbar. I figure it's more clear that `onError` is only called if the actual `mutation` fails

TBH I was originally not going to add the `onSuccess` callback at all, but I think the above distinction justifies it's usefulness. It's also consistent with other popular data-fetching libraries. LMK what you think

Reply:
Makes sense!